processing/src/main/java/io/druid/segment/filter/Filters.java

```diff
@@ -267,55 +291,160 @@ static double estimatePredicateSelectivity(
   }
 
   static double estimatePredicateSelectivity(
+      BitmapIndex bitmapIndex,
       ColumnSelector columnSelector,
       String dimension,
-      Iterable<ImmutableBitmap> bitmaps,
+      Iterable<Integer> bitmapIndexes,
       long totalNumRows
   )
   {
     final ColumnCapabilities columnCapabilities = columnSelector.getColumn(dimension).getCapabilities();
-    return estimatePredicateSelectivity(
-        bitmaps,
+    return estimateSelectivityOfBitmapList(
+        bitmapIndex,
+        bitmapIndexes,
         totalNumRows,
+        // assume multi-value column if columnCapabilities is null
         columnCapabilities == null || columnCapabilities.hasMultipleValues()
     );
   }
 
-  static double estimatePredicateSelectivity(
+  private static double estimateSelectivityOfBitmapList(
+      BitmapIndex bitmapIndex,
+      Iterable<Integer> bitmapIndexeIterable,
+      long totalNumRows,
+      boolean isMultiValueDimension
+  )
+  {
+    final List<Integer> bitmapIndexes = ImmutableList.copyOf(bitmapIndexeIterable);
+    long numMatchedRows = 0;
+    for (Integer index : bitmapIndexes) {
+      final ImmutableBitmap bitmap = bitmapIndex.getBitmap(index);
+      numMatchedRows += bitmap.size();
```

Comment:
Unfortunately Concise format doesn't store the size, so `size()` is not cheap. Roaring storing sizes of the blocks, but computing the total size of Roaring bitmap also requires some iteration and touching a lot of memory.

This problem also complicates implementation of #3878.

I think we should store bitmap sizes on segment format level and add ways to read bitmap size without creating bitmap itself.